**Data set from https://www.kaggle.com/competitions/playground-series-s3e14**

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
train_data = pd.read_csv("train.csv")

In [3]:
test_data = pd.read_csv("test.csv")

In [4]:
train_data.head()

id  clonesize  honeybee  bumbles  andrena  osmia  MaxOfUpperTRange  \
0   0       25.0      0.50     0.25     0.75   0.50              69.7   
1   1       25.0      0.50     0.25     0.50   0.50              69.7   
2   2       12.5      0.25     0.25     0.63   0.63              86.0   
3   3       12.5      0.25     0.25     0.63   0.50              77.4   
4   4       25.0      0.50     0.25     0.63   0.63              77.4   

   MinOfUpperTRange  AverageOfUpperTRange  MaxOfLowerTRange  MinOfLowerTRange  \
0              42.1                  58.2              50.2              24.3   
1              42.1                  58.2              50.2              24.3   
2              52.0                  71.9              62.0              30.0   
3              46.8                  64.7              55.8              27.0   
4              46.8                  64.7              55.8              27.0   

   AverageOfLowerTRange  RainingDays  AverageRainingDays  fruitset  fruitmass  \
0                  41.2         24.0                0.39  0.425011   0.417545   
1                  41.2         24.0                0.39  0.444908   0.422051   
2                  50.8         24.0                0.39  0.552927   0.470853   
3                  45.8         24.0                0.39  0.565976   0.478137   
4                  45.8         24.0                0.39  0.579677   0.494165   

       seeds       yield  
0  32.460887  4476.81146  
1  33.858317  5548.12201  
2  38.341781  6869.77760  
3  39.467561  6880.77590  
4  40.484512  7479.93417

In [5]:
X = train_data.drop(columns = ["id","yield"],axis = 1)
y = train_data["yield"]

In [6]:
Xt = test_data.drop(columns = ["id"],axis = 1)

In [7]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15289 entries, 0 to 15288
Data columns (total 16 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   clonesize             15289 non-null  float64
 1   honeybee              15289 non-null  float64
 2   bumbles               15289 non-null  float64
 3   andrena               15289 non-null  float64
 4   osmia                 15289 non-null  float64
 5   MaxOfUpperTRange      15289 non-null  float64
 6   MinOfUpperTRange      15289 non-null  float64
 7   AverageOfUpperTRange  15289 non-null  float64
 8   MaxOfLowerTRange      15289 non-null  float64
 9   MinOfLowerTRange      15289 non-null  float64
 10  AverageOfLowerTRange  15289 non-null  float64
 11  RainingDays           15289 non-null  float64
 12  AverageRainingDays    15289 non-null  float64
 13  fruitset              15289 non-null  float64
 14  fruitmass             15289 non-null  float64
 15  seeds              

In [8]:
X = X.dropna()

In [9]:
X.head()

clonesize  honeybee  bumbles  andrena  osmia  MaxOfUpperTRange  \
0       25.0      0.50     0.25     0.75   0.50              69.7   
1       25.0      0.50     0.25     0.50   0.50              69.7   
2       12.5      0.25     0.25     0.63   0.63              86.0   
3       12.5      0.25     0.25     0.63   0.50              77.4   
4       25.0      0.50     0.25     0.63   0.63              77.4   

   MinOfUpperTRange  AverageOfUpperTRange  MaxOfLowerTRange  MinOfLowerTRange  \
0              42.1                  58.2              50.2              24.3   
1              42.1                  58.2              50.2              24.3   
2              52.0                  71.9              62.0              30.0   
3              46.8                  64.7              55.8              27.0   
4              46.8                  64.7              55.8              27.0   

   AverageOfLowerTRange  RainingDays  AverageRainingDays  fruitset  fruitmass  \
0                  41.2         24.0                0.39  0.425011   0.417545   
1                  41.2         24.0                0.39  0.444908   0.422051   
2                  50.8         24.0                0.39  0.552927   0.470853   
3                  45.8         24.0                0.39  0.565976   0.478137   
4                  45.8         24.0                0.39  0.579677   0.494165   

       seeds  
0  32.460887  
1  33.858317  
2  38.341781  
3  39.467561  
4  40.484512

In [10]:
X.corr()

clonesize  honeybee   bumbles   andrena     osmia  \
clonesize              1.000000  0.304130  0.080433  0.065131 -0.007607   
honeybee               0.304130  1.000000 -0.017937  0.030671 -0.010394   
bumbles                0.080433 -0.017937  1.000000 -0.164962  0.158001   
andrena                0.065131  0.030671 -0.164962  1.000000  0.309556   
osmia                 -0.007607 -0.010394  0.158001  0.309556  1.000000   
MaxOfUpperTRange       0.016159  0.005840 -0.002104 -0.013061 -0.031391   
MinOfUpperTRange       0.015838  0.005755 -0.001813 -0.012928 -0.030819   
AverageOfUpperTRange   0.016057  0.005892 -0.001769 -0.012993 -0.031415   
MaxOfLowerTRange       0.016343  0.005942 -0.001613 -0.012924 -0.031398   
MinOfLowerTRange       0.016026  0.005809 -0.001804 -0.013035 -0.031486   
AverageOfLowerTRange   0.015987  0.005485 -0.001644 -0.013071 -0.031337   
RainingDays            0.165770  0.046494 -0.063294 -0.026572 -0.079874   
AverageRainingDays     0.164823  0.037532 -0.060232 -0.027193 -0.078720   
fruitset              -0.406793 -0.120492  0.160447  0.073669  0.209495   
fruitmass             -0.377688 -0.135310  0.163987  0.064722  0.192210   
seeds                 -0.396898 -0.139261  0.177022  0.063504  0.200597   

                      MaxOfUpperTRange  MinOfUpperTRange  \
clonesize                     0.016159          0.015838   
honeybee                      0.005840          0.005755   
bumbles                      -0.002104         -0.001813   
andrena                      -0.013061         -0.012928   
osmia                        -0.031391         -0.030819   
MaxOfUpperTRange              1.000000          0.998599   
MinOfUpperTRange              0.998599          1.000000   
AverageOfUpperTRange          0.999806          0.999004   
MaxOfLowerTRange              0.999503          0.998199   
MinOfLowerTRange              0.999829          0.998953   
AverageOfLowerTRange          0.999772          0.999040   
RainingDays                   0.011322          0.011727   
AverageRainingDays            0.010352          0.010767   
fruitset                      0.007580          0.008409   
fruitmass                     0.146237          0.147203   
seeds                         0.060963          0.061812   

                      AverageOfUpperTRange  MaxOfLowerTRange  \
clonesize                         0.016057          0.016343   
honeybee                          0.005892          0.005942   
bumbles                          -0.001769         -0.001613   
andrena                          -0.012993         -0.012924   
osmia                            -0.031415         -0.031398   
MaxOfUpperTRange                  0.999806          0.999503   
MinOfUpperTRange                  0.999004          0.998199   
AverageOfUpperTRange              1.000000          0.999465   
MaxOfLowerTRange                  0.999465          1.000000   
MinOfLowerTRange                  0.999973          0.999489   
AverageOfLowerTRange              0.999974          0.999423   
RainingDays                       0.011245          0.011302   
AverageRainingDays                0.010260          0.010262   
fruitset                          0.008503          0.007902   
fruitmass                         0.147676          0.146668   
seeds                             0.062082          0.061378   

                      MinOfLowerTRange  AverageOfLowerTRange  RainingDays  \
clonesize                     0.016026              0.015987     0.165770   
honeybee                      0.005809              0.005485     0.046494   
bumbles                      -0.001804             -0.001644    -0.063294   
andrena                      -0.013035             -0.013071    -0.026572   
osmia                        -0.031486             -0.031337    -0.079874   
MaxOfUpperTRange              0.999829              0.999772     0.011322   
MinOfUpperTRange              0.998953              0.999040     0.011727   
AverageOfUpperTRange

In [11]:
X.drop(columns = ["MinOfUpperTRange","MinOfUpperTRange","MaxOfLowerTRange","MinOfLowerTRange","RainingDays","MaxOfUpperTRange"],axis = 1,inplace=True)

In [12]:
Xt.drop(columns = ["MinOfUpperTRange","MinOfUpperTRange","MaxOfLowerTRange","MinOfLowerTRange","RainingDays","MaxOfUpperTRange"],axis = 1,inplace=True)

In [13]:
X.head()

clonesize  honeybee  bumbles  andrena  osmia  AverageOfUpperTRange  \
0       25.0      0.50     0.25     0.75   0.50                  58.2   
1       25.0      0.50     0.25     0.50   0.50                  58.2   
2       12.5      0.25     0.25     0.63   0.63                  71.9   
3       12.5      0.25     0.25     0.63   0.50                  64.7   
4       25.0      0.50     0.25     0.63   0.63                  64.7   

   AverageOfLowerTRange  AverageRainingDays  fruitset  fruitmass      seeds  
0                  41.2                0.39  0.425011   0.417545  32.460887  
1                  41.2                0.39  0.444908   0.422051  33.858317  
2                  50.8                0.39  0.552927   0.470853  38.341781  
3                  45.8                0.39  0.565976   0.478137  39.467561  
4                  45.8                0.39  0.579677   0.494165  40.484512

In [14]:
Xt.head()

clonesize  honeybee  bumbles  andrena  osmia  AverageOfUpperTRange  \
0       25.0      0.25     0.25     0.25   0.25                  71.9   
1       12.5      0.25     0.25     0.75   0.63                  79.0   
2       12.5      0.25     0.25     0.63   0.63                  71.9   
3       25.0      0.50     0.38     0.38   0.63                  71.9   
4       37.5      0.75     0.25     0.25   0.25                  79.0   

   AverageOfLowerTRange  AverageRainingDays  fruitset  fruitmass      seeds  
0                  50.8                0.39  0.399367   0.408088  31.394569  
1                  55.9                0.10  0.488048   0.442866  36.846956  
2                  50.8                0.26  0.583379   0.487057  40.037644  
3                  50.8                0.26  0.433014   0.422847  33.116091  
4                  55.9                0.39  0.360996   0.388860  29.558019

In [15]:
X["NewAverage"] = X["AverageOfUpperTRange"] +  X["AverageOfLowerTRange"]

In [16]:
Xt["NewAverage"] = Xt["AverageOfUpperTRange"] +  Xt["AverageOfLowerTRange"]

In [17]:
X = X.drop(columns = ["AverageOfLowerTRange","AverageOfLowerTRange"])

In [18]:
Xt = Xt.drop(columns = ["AverageOfLowerTRange","AverageOfLowerTRange"])

In [19]:
X["clonesize"].value_counts()

25.0    8245
12.5    6717
37.5     265
20.0      56
10.0       4
40.0       2
Name: clonesize, dtype: int64

In [20]:
cols = X.columns

In [21]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=58)

from sklearn.preprocessing import MinMaxScaler
sc = MinMaxScaler(feature_range=(0,1))

X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

X_train = pd.DataFrame(data = X_train,columns=[cols])
X_test = pd.DataFrame(data = X_test,columns=[cols])

In [22]:
Xt = sc.transform(Xt)
Xt = pd.DataFrame(data = Xt,columns=[cols])

In [23]:
from sklearn.svm import SVR
svr = SVR(C=2500,max_iter=500,gamma=0.04)
svr.fit(X_train,y_train)
prediction = svr.predict(X_test)

from sklearn.metrics import mean_absolute_error
print(f"Mean Absolute Error = {str(mean_absolute_error(y_test,prediction))}")

C:\Users\caner\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\svm\_base.py:299: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(


Mean Absolute Error = 1081.9905318521223


In [36]:
from xgboost import XGBRegressor
xgb = XGBRegressor(n_estimators=425,max_depth=5,learning_rate=0.020, random_state=58)

xgb.fit(X_train,y_train)
predict_xgb = xgb.predict(X_test)
print(f"Mean Absolute Error = {str(mean_absolute_error(y_test,predict_xgb))}")

Mean Absolute Error = 350.2337239411481


In [26]:
import tensorflow as tf

In [27]:
ann_model = tf.keras.models.Sequential()

In [38]:
ann_model.add(tf.keras.layers.Dense(units = 64,activation="relu"))
ann_model.add(tf.keras.layers.Dropout((0.5)))

ann_model.add(tf.keras.layers.Dense(units = 64,activation="relu"))
ann_model.add(tf.keras.layers.Dropout((0.5)))
              

              
ann_model.add(tf.keras.layers.Dense(units = 1,activation="linear"))

ann_model.compile(optimizer="adam",loss="mean_squared_error",metrics=["mae"])
ann_model.fit(X_train,y_train,epochs=25,batch_size=32,validation_data=(X_test,y_test))
              

Epoch 1/25
383/383 [==============================] - 7s 9ms/step - loss: 9605400.0000 - mae: 2471.0029 - val_loss: 6310040.0000 - val_mae: 2202.0488
Epoch 2/25
383/383 [==============================] - 3s 8ms/step - loss: 5837205.0000 - mae: 1961.9243 - val_loss: 3659184.7500 - val_mae: 1635.7682
Epoch 3/25
383/383 [==============================] - 2s 6ms/step - loss: 5609185.0000 - mae: 1922.9825 - val_loss: 5141294.5000 - val_mae: 1961.0983
Epoch 4/25
383/383 [==============================] - 2s 6ms/step - loss: 5575914.0000 - mae: 1906.9126 - val_loss: 3505998.2500 - val_mae: 1599.1984
Epoch 5/25
383/383 [==============================] - 2s 6ms/step - loss: 5513740.0000 - mae: 1894.5428 - val_loss: 3481120.5000 - val_mae: 1593.1833
Epoch 6/25
383/383 [==============================] - 3s 7ms/step - loss: 5325349.0000 - mae: 1871.8689 - val_loss: 4606913.0000 - val_mae: 1847.8280
Epoch 7/25
383/383 [==============================] - 2s 6ms/step - loss: 5309329.5000 - mae: 1867.1

In [120]:
predict_xgb_2 = xgb.predict(Xt)

In [123]:
ids = test_data["id"].values

In [124]:
ids

array([15289, 15290, 15291, ..., 25480, 25481, 25482], dtype=int64)

In [125]:
df = pd.DataFrame({"id" : ids,"yield" : predict_xgb_2})
df.to_csv("submission.csv",index=False)

xgb = XGBRegressor(n_estimators=425,max_depth=5,learning_rate=0.02, random_state=58) ---> Best Model 350 mean absolute error 